drivers/can/can_numaker.c

```diff
@@ -0,0 +1,279 @@
+
+#define NUMAKER_PINCTRL_DEFINE(inst)                                              \
+	PINCTRL_DT_INST_DEFINE(inst);
+#define NUMAKER_PINCTRL_INIT(inst)                                                \
+	.pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(inst),
+
+#define CAN_NUMAKER_INIT(inst)                                                    \
+	NUMAKER_PINCTRL_DEFINE(inst);                                             \
+	CAN_MCAN_DT_INST_CALLBACKS_DEFINE(inst, can_numaker_cbs_##inst);	  \
+										  \
+	static void can_numaker_irq_config_func_##inst(const struct device *dev); \
+										  \
+	static const struct can_numaker_config can_numaker_config_##inst = {      \
+		.canfd_base = CAN_MCAN_DT_INST_MCAN_ADDR(inst),                   \
+		.mrba = CAN_MCAN_DT_INST_MRBA(inst),			          \
+		.mram = CAN_MCAN_DT_INST_MRAM_ADDR(inst),			  \
+		.reset = RESET_DT_SPEC_INST_GET(inst),                            \
+		.clk_modidx = DT_INST_CLOCKS_CELL(inst, clock_module_index),      \
+		.clk_src = DT_INST_CLOCKS_CELL(inst, clock_source),               \
+		.clk_div = DT_INST_CLOCKS_CELL(inst, clock_divider),              \
+		NUMAKER_CLKCTRL_DEV_INIT(inst)                                    \
+		.irq_config_func = can_numaker_irq_config_func_##inst,            \
+		NUMAKER_PINCTRL_INIT(inst)                                        \
+	};                                                                        \
+                                                                                  \
+	static const struct can_mcan_config can_mcan_config_##inst =              \
+		CAN_MCAN_DT_CONFIG_INST_GET(inst,                                 \
+			&can_numaker_config_##inst,                               \
+			&can_numaker_ops,		                          \
+			&can_numaker_cbs_##inst);	                          \
+                                                                                  \
+	static uint32_t can_numaker_data_##inst;                                  \
+                                                                                  \
+	static struct can_mcan_data can_mcan_data_##inst =                        \
+		CAN_MCAN_DATA_INITIALIZER(&can_numaker_data_ ## inst);            \
+                                                                                  \
+	CAN_DEVICE_DT_INST_DEFINE(inst,                                           \
+		&can_numaker_init,                                                \
+		NULL,                                                             \
+		&can_mcan_data_##inst,                                            \
+		&can_mcan_config_##inst,                                          \
+		POST_KERNEL,                                                      \
+		CONFIG_CAN_INIT_PRIORITY,                                         \
+		&can_numaker_driver_api);                                         \
+									          \
+	static void can_numaker_irq_config_func_##inst(const struct device *dev)  \
```

Comment:
You could move this before `can_numaker_config`, saves you from having to forward declare it.

Reply:
Yes, it could save the declaration.
Update at commit 764cfb094.